File: backend/src/main/java/com/fptgang/backend/mapper/AccountMapper.java
package com.fptgang.backend.mapper;

import com.fptgang.backend.api.model.AccountDto;
import com.fptgang.backend.model.Account;
import com.fptgang.backend.util.DateTimeUtil;
import org.springframework.stereotype.Component;

@Component
public class AccountMapper extends BaseMapper<AccountDto, Account> {

    @Override
    public Account toEntity(AccountDto dto) {
        if (dto == null) {
            return null;
        }

        return Account.builder()
                .accountId(dto.getAccountId())
                .email(dto.getEmail())
                .firstName(dto.getFirstName())
                .lastName(dto.getLastName())
                .password(dto.getPassword())
                .avatarUrl(dto.getAvatarUrl())
                .balance(dto.getBalance())
                .updateBalanceAt(DateTimeUtil.fromOffsetToLocal(dto.getUpdateBalanceAt()))
                .role(mapRoleAccount(dto.getRole()))
                .isVerified(dto.getIsVerified())
                .verifiedAt(DateTimeUtil.fromOffsetToLocal(dto.getVerifiedAt()))
                .isVisible(dto.getIsVisible())
                .createdAt(DateTimeUtil.fromOffsetToLocal(dto.getCreatedAt()))
                .updatedAt(DateTimeUtil.fromOffsetToLocal(dto.getUpdatedAt()))
                .build();
    }

    @Override
    public AccountDto toDTO(Account entity, DetailLevel level) {
        if (entity == null) {
            return null;
        }

        AccountDto dto = new AccountDto();
        dto.setAccountId(entity.getAccountId());
        dto.setFirstName(entity.getFirstName());
        dto.setLastName(entity.getLastName());
        dto.setAvatarUrl(entity.getAvatarUrl());
        dto.setIsVisible(entity.getIsVisible());

        if (level == DetailLevel.REFERENCE) {
            return dto; // those fields are enough
        }

        dto.setEmail(entity.getEmail());
        dto.setBalance(entity.getBalance());
        //dto.setPassword(entity.getPassword()); // DO NOT RETURN PASSWORD
        dto.setUpdateBalanceAt(DateTimeUtil.fromLocalToOffset(entity.getUpdateBalanceAt()));
        dto.setRole(mapRoleAccountDto(entity.getRole()));
        dto.setIsVerified(entity.isVerified());
        dto.setVerifiedAt(DateTimeUtil.fromLocalToOffset(entity.getVerifiedAt()));
        dto.setCreatedAt(DateTimeUtil.fromLocalToOffset(entity.getCreatedAt()));
        dto.setUpdatedAt(DateTimeUtil.fromLocalToOffset(entity.getUpdatedAt()));

        return dto;
    }

    public Account.Role mapRoleAccount(AccountDto.RoleEnum roleEnum) {
        if (roleEnum == null) {
            return null; // Or a default Role, e.g., Role.CUSTOMER
        }

        switch (roleEnum) {
            case ADMIN:
                return Account.Role.ADMIN;
            case STAFF:
                return Account.Role.STAFF;
            case CUSTOMER:
                return Account.Role.CUSTOMER;
            default:
                throw new IllegalArgumentException("Unknown RoleEnum: " + roleEnum);
        }
    }

    public AccountDto.RoleEnum mapRoleAccountDto(Account.Role roleEnum) {
        if (roleEnum == null) {
            return null; // Or a default Role, e.g., Role.CUSTOMER
        }
        switch (roleEnum) {
            case ADMIN:
                return AccountDto.RoleEnum.ADMIN;
            case STAFF:
                return AccountDto.RoleEnum.STAFF;
            case CUSTOMER:
                return AccountDto.RoleEnum.CUSTOMER;
            default:
                throw new IllegalArgumentException("Unknown RoleEnum: " + roleEnum);
        }
    }

}

File: mobile/lib/data/mapper/account_mapper.dart
import 'package:mobile/data/mapper/shipping_info_mapper.dart';
import 'package:mobile/data/models/account_model.dart';
import 'package:openapi/api.dart';

import '../models/shipping_info_model.dart';

/// Maps between AccountDto (API data) and AccountModel (app domain model)
class AccountMapper {
  /// Converts an API DTO to the internal model
  static AccountModel toModel(AccountDto dto) {
    return AccountModel(
      accountId: dto.accountId,
      firstName: dto.firstName,
      lastName: dto.lastName,
      email: dto.email,
      password: dto.password,
      avatarUrl: dto.avatarUrl,
      balance: dto.balance ?? 0,
      updateBalanceAt: dto.updateBalanceAt,
      role: dto.role,
      isVerified: dto.isVerified ?? false,
      verifiedAt: dto.verifiedAt,
      isVisible: dto.isVisible ?? true,
      createdAt: dto.createdAt,
      updatedAt: dto.updatedAt,
      defaultShippingInfo: dto.defaultShippingInfo != null
          ? ShippingInfoMapper.toModel(dto.defaultShippingInfo!)
          : null,
    );
  }

  /// Converts an internal model to the API DTO
  static AccountDto toDto(AccountModel model) {
    return AccountDto(
      accountId: model.accountId,
      firstName: model.firstName,
      lastName: model.lastName,
      email: model.email,
      password: model.password,
      avatarUrl: model.avatarUrl,
      balance: model.balance,
      updateBalanceAt: model.updateBalanceAt,
      role: model.role,
      isVerified: model.isVerified,
      verifiedAt: model.verifiedAt,
      isVisible: model.isVisible,
      createdAt: model.createdAt,
      updatedAt: model.updatedAt,
      defaultShippingInfo: model.defaultShippingInfo != null
          ? ShippingInfoMapper.toDto(model.defaultShippingInfo!)
          : null,
    );
  }
}
